### ISSUE TITLE
Compilation failure due to missing symbol pthread_atfork

### ISSUE THREAD

**Issue by Flamefire** [scipy/scipy#11323]
#10614 introduced usage of `pthread_atfork` if the pthread header was found. However this requires adding `-lpthread` or `-pthread` to the compiler flags so the symbols can be found.

This kinda works on x86 platforms because Python is usually build with pthreads already and the dynamic library contains all the symbols. However on e.g. Power9 `libpthread` is a linker script contained a static part which contains e.g. `pthread_atfork` which does now not get linked and hence usage fails with 

  ```
File "/tmp/easybuild-tmp/eb-QqGzOJ/tmp81X11X/lib/python3.7/site-packages/scipy/__init__.py", line 156, in <module>
    from . import fft
  File "/tmp/easybuild-tmp/eb-QqGzOJ/tmp81X11X/lib/python3.7/site-packages/scipy/fft/__init__.py", line 81, in <module>
    from ._helper import next_fast_len
  File "/tmp/easybuild-tmp/eb-QqGzOJ/tmp81X11X/lib/python3.7/site-packages/scipy/fft/_helper.py", line 4, in <module>
    from . import _pocketfft
  File "/tmp/easybuild-tmp/eb-QqGzOJ/tmp81X11X/lib/python3.7/site-packages/scipy/fft/_pocketfft/__init__.py", line 3, in <module>
    from .basic import *
  File "/tmp/easybuild-tmp/eb-QqGzOJ/tmp81X11X/lib/python3.7/site-packages/scipy/fft/_pocketfft/basic.py", line 8, in <module>
    from . import pypocketfft as pfft
ImportError: /tmp/easybuild-tmp/eb-QqGzOJ/tmp81X11X/lib/python3.7/site-packages/scipy/fft/_pocketfft/pypocketfft.cpython-37m-powerpc64le-linux-gnu.so: undefined symbol: pthread_atfork
 (at easybuild/tools/run.py:529 in parse_cmd_output)
```

Proposed solution: Add `-pthread` or `-lpthread` to the link flags.

**Comment by larsoner** [comment]
@peterbell10 thoughts on this one?

**Comment by peterbell10** [comment]
>This kinda works on x86 platforms because Python is usually build with pthreads already and the dynamic library contains all the symbols.

It works on current linux builds because `distutils` adds `-pthread` to every extension build. Perhaps your python distribution is configured with another threading library? 

@Flamefire can you try building with gh-11324 to see if that fixes it.

**Comment by Flamefire** [comment]
I thought so too, but it isn't. To the best of my understanding distutils itself does not add `-pthread` as-is. It does (I assume) add the compile flags used to compile Python itself. I haven't dug through the logic how Python does determine the pthread flag, but it might be, that no flag is required for the shared library or `-lpthread` is used which is then not passed to extensions.

The reason why it works is actually NOT (or not only) that `-pthread` is added but that it is not required. Every extension links against libpython which is linked against lpthread so they get that "for free". However this is "by accident". Some platforms (e.g. power9) require the `-lpthread`.

**Comment by peterbell10** [comment]
>The reason why it works is actually NOT (or not only) that -pthread is added but that it is not required.

I don't understand you point here. If the extensions are compiled and linked with `-pthread` then they are linked with `-lpthread` no matter what, right?

**Comment by pv** [comment]
The build log contains the commands that were actualyl used, so uploading it here would clarify what flags were used.

**Comment by Flamefire** [comment]
> I don't understand you point here. If the extensions are compiled and linked with -pthread then they are linked with -lpthread no matter what, right?

This is correct. Let me rephrase: "The reason why it works is actually that -pthread is usually not required (and not that it is added by distutils which it isn't, at least not always). "

> The build log contains the commands that were actualyl used, so uploading it here would clarify what flags were used.

From scipy? Double checked and no pthread in the command line.

**Comment by Flamefire** [comment]
About Python: We are using EasyBuild to install it and that sets the environment variable `LIBS=-lpthread`, it might be that Python picks this up but distutils doesn't.

**Comment by peterbell10** [comment]
Would you mind posting the whole log?

**Comment by Flamefire** [comment]
I currently don't have access to it as our filesystem crashed. Can do tomorrow hopefully :/

But from what I could gather this is what happens:

- `export LIBS="-lpthread"`
- `cd python && ./configure [...] && make install`
  - autotools thinks no flag is required for pthread because it uses `LIBS`: 
- `cd scipy && pip install .` (or so)
  - pthread.h is found -> Pthread support enabled
  - distutils adds flags from python build -> no pthread flag (see above)
  - fft.so linked against `libpython.so` which is linked against `libpthread.so`
  - loading `fft.so` finds the loaded `libpthread.so` and uses it for pthread symbols
  - on x86 this is it and all works. on Power `libpthread.so` does not contain `pthread_atfork` which is in `libpthread.a`

`-lpthread` on power resolves to a linker script which links in the static and the dynamic library parts.

**Comment by Flamefire** [comment]
@peterbell10 As the logs are gone and recreating them takes a couple hours: Do you still need the logs? IMO my analysis above is complete and #11324 fixes the problem so I don't know of what help the logs would be.

**Comment by peterbell10** [comment]
That's find @Flamefire. I think we've adequately diagnosed the problem without the log.
